File: src/main/java/net/orandja/vw/mods/ProtectBlock/ProtectBlock.java
package net.orandja.vw.mods.ProtectBlock;

import net.minecraft.entity.Entity;
import net.minecraft.util.math.BlockPos;
import net.minecraft.world.World;
import org.apache.commons.compress.utils.Lists;
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;

import java.util.List;

public interface ProtectBlock {

    interface PredicateBlockWithEntity {
        boolean test(World world, BlockPos pos, Entity entity);
    }

    interface PredicateBlock {
        boolean test(World world, BlockPos pos);
    }

    // returns true if the player can destroy the block
    List<PredicateBlockWithEntity> DESTROY = Lists.newArrayList();
    // returns true if the explosion can destroy the block
    List<PredicateBlock> EXPLOSION_PROTECTION = Lists.newArrayList();
    // returns false if the hopper can extract item from block **Might be changed in the future
    List<PredicateBlock> EXTRACTION_PREVENTION = Lists.newArrayList();

    World getWorld();
    List<BlockPos> getAffectedBlocks();
    void setAffectedBlocks(List<BlockPos> list);

    default boolean preventsExtract(World world, BlockPos pos) {
        return EXTRACTION_PREVENTION.stream().anyMatch(predicate -> predicate.test(world, pos));
    }

    default boolean preventsExtract(World world, double x, double y, double z) {
        return preventsExtract(world, new BlockPos(x, y, z));
    }

    default boolean canExplode(World world, BlockPos pos) {
        return EXPLOSION_PROTECTION.stream().noneMatch(predicate -> predicate.test(world, pos));
    }

    default void onExplosion() {
        List<BlockPos> newBlocks = getAffectedBlocks().stream().filter(block -> getWorld() != null && canExplode(getWorld(), block)).toList();
        getAffectedBlocks().clear();
        getAffectedBlocks().addAll(newBlocks);
    }

    default boolean canDestroy(World world, BlockPos pos, Object entity) {
        if(entity instanceof Entity entity1) {
            return DESTROY.stream().anyMatch(predicate -> predicate.test(world, pos, entity1));
        }

        return true;
    }

    default void onPlayerDestroy(World world, BlockPos pos, Object entity, CallbackInfoReturnable<Boolean> info) {
        if(!canDestroy(world, pos, this)) {
            info.setReturnValue(true);
        }
    }

}

File: src/main/kotlin/net/orandja/vw/logic/ProtectBlock.kt
package net.orandja.vw.logic

import net.minecraft.entity.Entity
import net.minecraft.util.math.BlockPos
import net.minecraft.world.World
import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable

interface ProtectBlock {

    companion object {

        // returns true if the player can destroy the block
        val DESTROY: ArrayList<(world: World, pos: BlockPos, entity: Entity) -> Boolean> = ArrayList()
        // returns true if the explosion can destroy the block
        val EXPLOSION_PROTECTION: ArrayList<(world: World, pos: BlockPos) -> Boolean> = ArrayList()
        // returns false if the hopper can extract item from block **Might be changed in the future
        val EXTRACTION_PREVENTION: ArrayList<(world: World, pos: BlockPos) -> Boolean> = ArrayList()

    }

    fun preventsExtract(world: World, pos: BlockPos): Boolean {
        return EXTRACTION_PREVENTION.find { it(world, pos) } != null
    }

    fun preventsExtract(world: World, x: Double, y: Double, z: Double): Boolean {
        return preventsExtract(world, BlockPos(x, y, z))
    }

    val world: World?
    @Suppress("PLATFORM_CLASS_MAPPED_TO_KOTLIN")
    var affectedBlocks: java.util.List<BlockPos>

    fun canExplode(world: World, pos: BlockPos): Boolean {
        return EXPLOSION_PROTECTION.find { it(world, pos) } == null
    }

    fun onExplosion() {
        val newBlocks = affectedBlocks.filter { world != null && canExplode(world!!, it) }.toList()
        affectedBlocks.clear()
        affectedBlocks.addAll(newBlocks)
    }

    fun canDestroy(world: World, pos: BlockPos, player: Any): Boolean {
        return DESTROY.find { it(world, pos, player as Entity) } != null
    }

    fun onPlayerDestroy(world: World, pos: BlockPos, player: Any, info: CallbackInfoReturnable<Boolean>) {
        if (!canDestroy(world, pos, this))
            info.returnValue = true;
    }
}